src/ucp/tag/eager_rcv.c

```diff
@@ -70,27 +94,33 @@ ucp_eager_handler(void *arg, void *data, size_t length, void *desc,
             /* TODO In case an error status is returned from ucp_tag_process_recv,
              * need to discard the rest of the messages */
             status = UCS_OK;
+
+            if (flags & UCP_RECV_DESC_FLAG_SYNC) {
+                ucp_eager_sync_send_handler(arg, data, flags);
+            }
+
             goto out;
         }
     }
 
+    rdesc = ucp_recv_desc_get(worker, data, desc, length, hdr_len, flags);
+    if (!rdesc) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err;
+    }
+
     ucs_trace_req("unexp recv %c%c%c tag %"PRIx64" length %zu desc %p",
                   (flags & UCP_RECV_DESC_FLAG_FIRST) ? 'f' : '-',
                   (flags & UCP_RECV_DESC_FLAG_LAST)  ? 'l' : '-',
                   (flags & UCP_RECV_DESC_FLAG_EAGER) ? 'e' : '-',
                   recv_tag, length, rdesc);
 
-    if (data != rdesc + 1) {
-        memcpy(rdesc + 1, data, length);
-    }
-
-    rdesc->length  = length;
-    rdesc->hdr_len = hdr_len;
-    rdesc->flags   = flags;
     ucs_queue_push(&context->tag.unexpected, &rdesc->queue);
 
-    status = UCS_INPROGRESS;
 out:
+    status = (rdesc && (data == rdesc + 1)) ? UCS_INPROGRESS : UCS_OK;
```

Comment:
again - need to avoid the extra overhead, even if same end result, since this is fast path 
see https://github.com/openucx/ucx/blob/master/doc/OptimizationStyle.md